### ISSUE TITLE
IsItemHovered returns true while it is not

### ISSUE THREAD

**Issue by wolfgang371** [ocornut/imgui#3851]
Hi Omar,

I have cut my problem down to the snippet below.
If you approach the splitter from above, it's fine.
If you approach from below, the mouse cursor changes one pixel too early, without the splitter getting green. (BTW: not part of the code: the splitter can only be activated if it's green).
I also tried that with a vertical splitter bar, there everything is fine.
In my real application it's also two pixel, not just one, so probably something stacks there...

(this is Crystal, I hope that's not a problem...)

```crystal
class OffByOne
    def paint
        size = 200.0f32
        draw_splitter(size)
        ImGui.begin_child("##a", ImVec2.new(0, size))
        ImGui.text("foo")
        ImGui.end_child
        ImGui.begin_child("##b", ImVec2.new(0, size))
        ImGui.text("bar")
        ImGui.end_child
    end
    def draw_splitter(size : Float32)
        backup_pos = ImGui.get_cursor_pos
        ImGui.set_cursor_pos_y(backup_pos.y + size)
        ImGui.push_style_color(ImGuiCol::Button, ImGui.col32(130,130,130,255))
        ImGui.push_style_color(ImGuiCol::ButtonHovered, ImGui.col32(0,255,0,255))
        ImGui.button("##splitter", ImVec2.new(-1.0f32, 5.0f32))
        if ImGui.is_item_hovered
            # this one seems to show up one pixel too early if you approach the splitter from _below_
            ImGui.set_mouse_cursor(ImGuiMouseCursor::ResizeNS)
        end
        ImGui.pop_style_color(2)
        ImGui.set_cursor_pos(backup_pos)
    end
end
```

```
Dear ImGui 1.80 (18000)
--------------------------------
sizeof(size_t): 8, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=201402
define: __linux__
define: __GNUC__=7
--------------------------------
io.BackendPlatformName: imgui_impl_sfml
io.BackendRendererName: NULL
io.ConfigFlags: 0x00000000
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x00000007
 HasGamepad
 HasMouseCursors
 HasSetMousePos
--------------------------------
io.Fonts: 1 fonts, Flags: 0x00000000, TexSize: 512,64
io.DisplaySize: 1280.00,720.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 8.00,8.00
style.WindowBorderSize: 1.00
style.FramePadding: 4.00,3.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 8.00,4.00
style.ItemInnerSpacing: 4.00,4.00
```

**Comment by rokups** [comment]
This is a very interesting edge-case. I think you are using API incorrectly and this is also may be a bug, @ocornut would have a better idea regarding bug part.

What happens here is that you first render a splitter button in the middle of window parent window. Then you render two child windows on top of splitter button. See in immediate mode UIs function call order is important. So now when you hover a button `ItemHoverable()` still fails on that stray pixel because hovered window (which is child `##b`) does not match window which splitter button belongs to.

What you should be doing is something like this:
```crystal
class OffByOne
    size = 200.0f32

    def paint
        ImGui.begin_child("##a", ImVec2.new(0, size))
        ImGui.text("foo")
        ImGui.end_child
        draw_splitter(size)
        ImGui.begin_child("##b", ImVec2.new(0, ImGui.GetContentRegionMax.y - size))       # !!!
        ImGui.text("bar")
        ImGui.end_child
    end
    def draw_splitter(size : Float32)
        ImGui.push_style_color(ImGuiCol::Button, ImGui.col32(130,130,130,255))
        ImGui.push_style_color(ImGuiCol::ButtonHovered, ImGui.col32(0,255,0,255))
        ImGui.button("##splitter", ImVec2.new(-1.0f32, 5.0f32))
        if ImGui.is_item_active
            # this one seems to show up one pixel too early if you approach the splitter from _below_
            ImGui.set_mouse_cursor(ImGuiMouseCursor::ResizeNS)

            # !!!
            prev_size = size
            size += (int)ImGui.GetMouseDragDelta.y
            size = size < 0 ? 0 : size
            if (size != prev_size)
                ImGui.ResetMouseDragDelta

        end
        ImGui.pop_style_color(2)
    end
end
```

This is not complete, you would need to take into account all kinds of paddings (window pading, item spacing) as well to get rid of the vertical scrollbar, but this gives you idea i hope. What i did here is i placed splitter button between child windows so that no children overlaps with a splitter. This solves the issue with `ItemHoverable()` getting confused.

**Comment by ocornut** [comment]
I don't think the submission order meaningfully matters here. The problem is that `IsItemHovered()` essentially compares `g.HoveredRootWindow != window->RootWindow` whereas `ItemHoverable()` compares `g.HoveredWindow != window`.

C++ repro
```cpp
void DrawSplitter(float size)
{
    ImVec2 backup_pos = ImGui::GetCursorPos();
    ImGui::SetCursorPosY(backup_pos.y + size);
    ImGui::PushStyleColor(ImGuiCol_Button, IM_COL32(130, 130, 130, 255));
    ImGui::PushStyleColor(ImGuiCol_ButtonHovered, IM_COL32(0, 255, 0, 255));
    if (ImGui::GetIO().KeyShift)// breakpoint
        ImGui::GetIO();
    ImGui::Button("##splitter", ImVec2(-1.0f, 5.0f));
    if (ImGui::IsItemHovered())
        ImGui::SetMouseCursor(ImGuiMouseCursor_ResizeNS);
    ImGui::PopStyleColor(2);
    ImGui::SetCursorPos(backup_pos);
}
```

```cpp
ImGui::Begin("Test 3851");
float size = 200.0f;
DrawSplitter(200.0f);

ImGui::BeginChild("##a", ImVec2(0, size));
ImGui::Text("foo");
ImGui::EndChild();
ImGui::BeginChild("##b", ImVec2(0, size));
ImGui::Text("bar");
ImGui::EndChild();

ImGui::End();
```

I think `IsItemHovered()` is faulty:

```cpp
// Test if we are hovering the right window (our window could be behind another window)
// [2017/10/16] Reverted commit 344d48be3 and testing RootWindow instead. I believe it is correct to NOT test for RootWindow but this leaves us unable to use IsItemHovered() after EndChild() itself.
// Until a solution is found I believe reverting to the test from 2017/09/27 is safe since this was the test that has been running for a long while.
//if (g.HoveredWindow != window)
//    return false;
if (g.HoveredRootWindow != window->RootWindow && !(flags & ImGuiHoveredFlags_AllowWhenOverlapped))
    return false;
```
I believe 344d48be3 was right and that comment that came with the revert (c76f01429, for #1370) is particularly dubious.

Need to investigate.

**Comment by ocornut** [comment]
The revert c76f014 for #1370 state this was meant to be temporary:
> "I don't have a satisfying solution just yet, so for now I have reverted that changes and added notes about it. So this is fixed here."

Spoiler: it stayed like that for 3 years.

Working on a solution now.

**Comment by ocornut** [comment]
Pushed a fix ee643b2ad919f9e85b973f3915d13a4fc0612a5a

For reference this is what I'm testing right now, may add more in the future:

```cpp
// ## Test for IsItemHovered() on BeginChild() and InputMultiLine() (#1370, #3851)
t = IM_REGISTER_TEST(e, "widgets", "widgets_hover");
t->GuiFunc = [](ImGuiTestContext* ctx)
{
    ImGuiTestGenericVars& vars = ctx->GenericVars;
    ImGui::Begin("Test Window", NULL, ImGuiWindowFlags_NoSavedSettings);

    ImGui::BeginChild("Child", ImVec2(100, 100), true);
    vars.Pos = ImGui::GetCursorScreenPos();
    ImGui::Dummy(ImVec2(10, 10));
    ImGui::Button("button");
    vars.Id = ImGui::GetItemID();
    ImGui::EndChild();
    vars.Bool1 = ImGui::IsItemHovered();
    ImGui::Text("hovered: %d", vars.Bool1);

    ImGui::InputTextMultiline("##Field", vars.Str1, IM_ARRAYSIZE(vars.Str1), ImVec2(-FLT_MIN, 0.0f));
    vars.Bool2 = ImGui::IsItemHovered();
    ImGui::Text("hovered: %d", vars.Bool2);
    ImGui::End();
};
t->TestFunc = [](ImGuiTestContext* ctx)
{
    ImGuiTestGenericVars& vars = ctx->GenericVars;
    ctx->SetRef("Test Window");

    // move on child's void
    ctx->MouseMoveToPos(vars.Pos);
    IM_CHECK(vars.Bool1 == true);
    IM_CHECK(vars.Bool2 == false);

    // move on child's item
    ctx->MouseMove(vars.Id);
    IM_CHECK(vars.Bool1 == true);
    IM_CHECK(vars.Bool2 == false);

    // move on InputTextMultiline() = child embedded in group
    ctx->MouseMove("##Field");
    IM_CHECK(vars.Bool1 == false);
    IM_CHECK(vars.Bool2 == true);
};
```

**Comment by wolfgang371** [comment]
that was quick - thank you! I can test it when there's a new version available for crystal
